packages/core/src/component.ts

```diff
@@ -15,6 +15,7 @@ export interface ProviderMap {
 export interface Component {
   controllers?: Constructor<any>[];
   providers?: ProviderMap;
+  [prop: string]: any;
```

Comment:
Repository is a first class citizen. We probably should have:
```
repositories?: Constructor<any>[];
```

Reply:
I'm not sure if `repositories` should be a first class citizen because `@loopback/repository` is an optional package. This way allows other extensions to specify their own properties on the Component should they need to do so. For `controllers` and `providers` the logic to bind them resides in core and hence it makes sense for them to be first class citizens. IMO if/when they are refactored out then even those shouldn't be first class citizens of the core. Only the properties used by the implementing package should be considered first class citizens.

Reply:
+1

IMO, juggler-backed `repository` should not be a first class citizen in loopback-next, we should encourage users to use their flavor of ORM/Repository.

We could introduce a generic `Repository` interface in `@loopback/core`, but I am concerned that in order to support all different ORMs out there, will will end up with an empty `Repository` interface, which feels like a ceremonial overhead. Yes, `app.repository` argument restricts the types allowed, but since it's a constructor of an empty interface, I think it will allow any class constructor, therefore rendering the type system/checks useless. (Remember, TypeScript is comparing interfaces using their properties - it considers two interfaces as the same if they have the same properties.)

> This is different from the following style:
> ` controller(controllerCtor: ControllerClass): Binding`

As I am thinking about the support for multiple transports (e.g. gRPC adapter - #521), I think we should remove the REST transport from `@loopback/core` into a standalone package/extension. When this is done, then it may make sense to move `app.controller` API into that extension too. (Depending on whether we expect a single controller to be accessed by multiple transports.)

Reply:
Is there a way how to avoid this change? My concern is that by allowing any property with any value, we are loosing type safety.

Reply:
@bajtos Would strongly-typing these properties mean that the underlying JavaScript will now be doing "type checking" to ensure that they're compliant? I think the difficulty here is that if we enforce a type here, it will force users that don't want to use TypeScript to jump through extra hoops to ensure all of their classes/prototype objects are properly extending some base object.

>IMO, juggler-backed repository should not be a first class citizen in loopback-next, we should encourage users to use their flavor of ORM/Repository.

I agree completely; the only parts of loopback-next I feel 100% certain in calling first-class citizens would be the Controllers and the Providers, because they're suitably restrictive in the sense that what they require of users seems to be fairly universal in its compatibility and appeal. If we had a solid definition of a Repository that wasn't empty, but was firm in its universal assertions of what _all_ repositories should logically contain, then I'd be all for it.

Reply:
As a side note, if we have an interface that meets those criteria, it _could_ also be a good foundation for relations and a DSL. We'd have to toy around more with the idea.

Reply:
> Would strongly-typing these properties mean that the underlying JavaScript will now be doing "type checking" to ensure that they're compliant? I think the difficulty here is that if we enforce a type here, it will force users that don't want to use TypeScript to jump through extra hoops to ensure all of their classes/prototype objects are properly extending some base object.

No no, this will be all at TypeScript compile-time only.

Here is what I have in mind: our repository mixin should somehow contribute `repositories` property to the `Component` type. At minimum we can cast `appConfig.components` to something like `Component & {repositories: Array<typeof Repository>}`. Similarly if we want `appConfig` to include repositories, then we should have a type definition for that.

For example:

```ts
export interface WithRepositories {
  repositories?: Array<typeof Repository>;
}

// in the mixin
constructor(...args: any[]) {
  super(...args);
  const config: AppConfig & WithRepositories = args[0];
  if (config.repositories) {
    // ...
  }
}

component(component: Component & WithRepositories) {
  super.component(component);
  if (component.repositories) {
    // ...
  }
}
```

Reply:
IMO, `controller` and `repository` are in the same nature. Both are LoopBack opinions/constructs. They should follow the same style to be registered with the application. I'm fine to treat both as extensions. IMO, almost every thing in LoopBack next are extensions, either built-in or contributed by other modules.

Reply:
I can create a ticket to refactor `controller` into an extension similar to repository. 

> Is there a way how to avoid this change? My concern is that by allowing any property with any value, we are loosing type safety.

One way would be to create an optional property called `repositories` which would make it a first class citizen as suggested by @raymondfeng ... however my bigger reason for this approach was so component extension developers can easily require additional properties and just document them in their component. As for breaking type safety, we have types for the properties required by core as optional properties so they are still typed and additional properties shouldn't be our concern since they shouldn't end up breaking anything.

Reply:
^ Also @bajtos if we don't allow random properties in Application / Component, we can't call `super()` with `...args` because it'll contain additional properties like `repositories` and will throw an error. 

We can also do something like you suggested in your example but that would require an importing `ApplicationOptions` from `@loopback/core` and hence creating a dependency on it & it still requires the `any` in Application. It does allow us to check the type of repository property in the mixin (which can be beneficial and sets a best practice example). If the dependency is ok I can update it accordingly.